test/gpu/native/cudaOnly/intents/foreachInIntentCustomIter.chpl

```diff
@@ -0,0 +1,63 @@
+use GpuDiagnostics;
+
+extern {
+  #include <cuda_runtime.h>
+
+  static __device__ __host__ void printIt(const char *msg, int idx) {
+    printf("%s: %i\n", msg, idx);
+  }
+}
+
+pragma "codegen for GPU"
+pragma "always resolve function"
+extern proc printIt(msg : c_ptrConst(c_char), idx : c_int) : void;
+
+
+iter myIter() {
+  var __CFL_IDX__ : int;
+  while __primitive("C for loop",
+          __primitive( "=", __CFL_IDX__, 0),
+          __primitive("<=", __CFL_IDX__, 0),
+          __primitive("+=", __CFL_IDX__, 1))
+  {
+    yield __CFL_IDX__;
+  }
+
+  while __primitive("C for loop",
+          __primitive( "=", __CFL_IDX__, 1),
+          __primitive("<=", __CFL_IDX__, 1),
+          __primitive("+=", __CFL_IDX__, 1))
+  {
+    yield __CFL_IDX__;
+  }
+}
+
+// --------------
+
+proc doit() {
+  writeln("ON CPU ---");
+  var __VAR__ = 100;
+  writeln("BEFORE ALL: ", __VAR__);
+  foreach __IDX__ in myIter() with (in __VAR__) {
+    printIt("BEFORE":c_ptrConst(c_char), __VAR__ : c_int);
+    __VAR__ = __IDX__;
+    printIt("AFTER":c_ptrConst(c_char), __VAR__ : c_int);
+  }
+  writeln("AFTER ALL: ", __VAR__);
+
+  writeln();
+
+  on here.gpus[0] {
+    writeln("ON GPU ---");
+    var __VAR__ = 100;
+    writeln("BEFORE ALL: ", __VAR__);
+    foreach __IDX__ in myIter() with (in __VAR__) {
+      printIt("BEFORE":c_ptrConst(c_char), __VAR__ : c_int);
+      __VAR__ = __IDX__;
+      printIt("AFTER":c_ptrConst(c_char), __VAR__ : c_int);
+    }
+    writeln("AFTER ALL: ", __VAR__);
+  }
+}
+
+doit();
```

Comment:
I am confused why the tests are `cudaOnly` and use `printIt`. Can't we do this on AMD and use `gpuWriteln`?

Reply:
Unfortunately no, `gpuWriteln` takes only a single `c_string` argument.

I could write an AMD specific version of the extern and rig up our test system to link as appropriate but that feels like jumping through hoops without much benefit.

Ideally, we should be figuring out how we could generalize `gpuWriteln`. Do you think it would be possible to have it take a variable number of arguments and then forward those to `printf`?

Reply:
Hmm, that's right. Can you 

```chpl
pragma "codegen for GPU"
extern proc printf(s...) : void;
```

and then use `printf` from inside your kernels?

In general, we can consider improving what we have in `gpuWriteln`, I agree. Or, we could remove it and replace it with `gpuPrintf` for the time being.

I was also thinking we could do `gpuWritef(param format: string, ...)`, but then again there are differences between C's `printf` and Chapel's `writef` formats, and that that could be confusing.

Reply:
If I change the Chapel `gpuWrite` proc to this:

```chpl
  inline proc gpuWrite(const args ...) {
    pragma "codegen for CPU and GPU"
    extern proc printf(args ...);
    printf((...args));
  }
```

And run `gpuWriteln("hello %i\n" : c_ptrConst(c_char), 42);` I'll get "ptxas /tmp/chpl-stonea.deleteme-rnGv3q/chpl__gpu.s, line 2555; error   : Call has wrong number of parameters".

If I change `chpl_gpu_write_ to have this signature

```chpl
MAYBE_GPU static inline void chpl_gpu_write(const char *str, ...)
```

Then I get this error:

"error: CUDA device code does not support variadic functions"

Reply:
I've added a slew of overloaded `gpuPrintf` functions to deal with this in the meantime.